src/qpx_scan_algo.cpp

```diff
@@ -0,0 +1,940 @@
+        buf[cnt - 1] = '\n';
+        buf[cnt] = 0;
+        return cnt;
+      }
+      cnt++;
+      cbuf++;
+    }
+  }
+  if (stop_req)
+    return -1;
+  buf[cnt] = '\n';
+  buf[cnt + 1] = 0;
+  return cnt + 1;
+}
+
+//
+// CD/DVD read transfer rate
+//
+
+int qscanner::run_rd_transfer() {
+  //	int     flushcnt = FLUSH_LINES;
+  bool use_readcd = 0;
+  int bsize = -1;
+  int rsize = -1;
+  long lba;
+  //	double	btime;
+  int err = 0;
+  int br = 0;
+  dev->parms.read_speed_kb = (int)(speed * dev->parms.speed_mult);
+  set_rw_speeds(dev);
+  get_rw_speeds(dev);
+  speed = (int)(dev->parms.read_speed_kb / (dev->parms.speed_mult - 0.5));
+  if (dev->media.type & DISC_CD) {
+    printf("Running READ transfer rate test on CD at speed %d...\n", speed);
+    if (dev->capabilities & CAP_DAE)
+      use_readcd = 1;
+    rsize = 15;
+    // bsize=75;
+    bsize = rsize * 10; // 2 sec
+    spd1X = 150;
+  } else if (dev->media.type & DISC_DVD) {
+    printf("Running READ transfer rate test on DVD at speed %d...\n", speed);
+    rsize = 16;
+    bsize = rsize * 64; // 2M
+    spd1X = 1385;
+    // check if DVD is region-protected
+    // dev->silent++;
+    get_rpc_state(dev);
+    read_disc_regions(dev);
+    // dev->media.dvdcss.method = DVDCSS_METHOD_DISC;
+    switch (dev->media.dvdcss.protection) {
+    case 0: // unprotected DVD
+      dev->media.dvdcss.method = DVDCSS_METHOD_NONE;
+      break;
+    case 1: // CSS/CPPM protected DVD
+    case 2: // CPRM - protected DVD
+      // just to auth to be able read data
+      dev->media.dvdcss.method = DVDCSS_METHOD_KEY;
+      if (css_disckey(dev)) {
+        printf("DVD auth failure!\n");
+        return -1;
+      }
+      break;
+    default:
+      printf("Unknown DVD protection scheme: %02X\n",
+             dev->media.dvdcss.protection);
+      dev->media.dvdcss.method = DVDCSS_METHOD_NONE;
+      break;
+    }
+    // dev->silent--;
+  } else if (dev->media.type & DISC_BD) {
+    printf("Running READ transfer rate test on DVD at speed %d...\n", speed);
+    rsize = 16;
+    bsize = rsize * 64; // 2M
+    spd1X = 4500;
+  } else {
+    printf("Can't run read transfer rate test: unsupported media!\n");
+    return -1;
+  }
+  if (stop_req)
+    return 0;
+  printf("Using %s command\n", use_readcd ? "READ CD" : "READ");
+  //	set_read_speed();
+  //	wait_unit_ready(dev, 6);
+
+  spinup(dev, 4); // spin up (4 seconds)
+
+  //	flush_cache(dev);
+  seek(dev, lba_sta);
+  //	wait_unit_ready(dev, 6);
+  msleep(300);
+  gettimeofday(&s, NULL);
+  printf("Reading blocks: %ld - %ld (%ld MB)\n", lba_sta, lba_end,
+         (lba_end - lba_sta) >> 9);
+  gettimeofday(&blks, NULL);
+#ifdef USE_FFLUSH
+  fflush(stdout);
+#endif
+  for (lba = lba_sta; (!stop_req) && !err && lba < lba_end; lba += rsize) {
+    if (lba + rsize >= lba_end)
+      rsize = lba_end - lba;
+    if (use_readcd)
+      err = read_cd(dev, dev->rd_buf, lba, rsize, 0xF8);
+    else
+      err = read(dev, dev->rd_buf, lba, rsize);
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)